src/index.ts

```diff
@@ -23,108 +39,203 @@ export {
+	readable?: Array<() => void>;
+	'resolved-name'?: Array<(dockerfilePath: string) => void>;
+	resolver?: Array<(resolverName: string) => void>;
+}
+
+/**
+ * Reads bundle.tarStream (typically a reference to BuildTask.buildStream) and
+ * resolves it to a Docker-compatible tar stream (i.e. selects a specified
+ * Dockerfile, translates Dockerfile.template vars and the like).
+ * @param bundle Bundle including the input tar stream
+ * @param resolvers Bundle resolvers
+ * @param resolveListeners Event listeners for tar stream resolution.
+ * Note that although you can add listeners to the returned Pack stream,
+ * the stream is piped to before it is returned and events may be missed
+ * unless the listeners are installed through the resolveListeners argument.
+ * You should always add at least an 'error' handler, or uncaught errors
+ * may crash the app.
+ * @param dockerfile User-selected Dockerfile (docker-compose `dockerfile:`)
+ */
 export function resolveInput(
 	bundle: Bundle,
 	resolvers: Resolver[],
+	resolveListeners: ResolveListeners,
 	dockerfile?: string,
-): Readable {
+): tar.Pack {
 	const extract = tar.extract();
 	const pack = tar.pack();
+	for (const event of Object.keys(resolveListeners) as Array<
+		keyof ResolveListeners
+	>) {
+		for (const listener of resolveListeners[event] || []) {
+			pack.on(event, listener);
+		}
+	}
 	let specifiedFileResolver: null | Resolver = null;
 	if (dockerfile != null) {
 		// Ensure that this will match the entry in the tar archive
 		dockerfile = Utils.normalizeTarEntry(dockerfile);
 	}
 
+	extract.on('error', (error: Error) => {
+		console.log('resin-bundle-resolve extract on error', error);
+		pack.emit('error', error);
+	});
+
 	extract.on(
 		'entry',
 		async (header: tar.Headers, stream: Readable, next: () => void) => {
-			const name = Utils.normalizeTarEntry(header.name);
+			try {
+				console.log('resin-bundle-resolve extract on entry');
 
-			// Handle the case where the dockerfile is specified
-			if (dockerfile != null) {
-				if (specifiedFileResolver == null && name === dockerfile) {
-					specifiedFileResolver = await resolveSpecifiedFile(
-						resolvers,
+				specifiedFileResolver =
+					(await resolveTarStreamOnEntry(
+						{ header, stream, next },
 						bundle,
-						name,
-						header,
-						stream,
+						resolvers,
 						pack,
-					);
-					next();
-					return;
-				}
-			} else {
-				const potentials = resolvers.filter(r => r.needsEntry(name));
-				if (potentials.length > 0) {
-					const entry = await streamToFileInfo(stream, header);
-					for (const resolver of potentials) {
-						resolver.entry(entry);
-					}
-					// Also add it to the stream
-					pack.entry(header, entry.contents);
-					next();
-					return;
-				}
+						specifiedFileResolver,
+						dockerfile,
+					)) || specifiedFileResolver;
+			} catch (error) {
+				console.log('resin-bundle-resolve extract entry -> error', error);
+				pack.emit('error', error);
```

Comment:
Some errors thrown in stream listeners were not being caught, so I added the try-catch above and "forwarded" the errors through `pack.emit('error')`.